src/transport_udp.cpp

```diff
@@ -1,237 +1,515 @@
+
+    std::lock_guard<std::mutex> _(_socket_write_mutex);
+
+    const auto data_len = sizeof(dhdr) + cd.data.size();
+    if (data_len > sizeof(data)) {
+        logger->error << "conn_id: " << conn.id << " data_len: " << data_len << " is too large" << std::flush;
+        return false;
+    }
+
+    memcpy(data, &dhdr, sizeof(dhdr));
+    memcpy(data + sizeof(dhdr), cd.data.data(), cd.data.size());
+
+
+    int numSent = sendto(fd,
+                         data,
+                         data_len,
+                         0 /*flags*/,
+                         (struct sockaddr *) &conn.addr.addr,
+                         conn.addr.addr_len);
+
+    if (numSent < 0) {
+        logger->error << "conn_id: " << conn.id
+                      << " Error sending DATA to UDP socket: " << strerror(errno)
+                      << std::flush;
+
+        return false;
+
+    } else if (numSent != data_len) {
+        logger->info << "conn_id: " << conn.id
+                     << " Failed to send DATA len: " << data_len << ", sent: " << numSent
+                     << std::flush;
+        return false;
+    }
+
+    conn.tx_report_metrics.total_bytes += cd.data.size();
+    conn.tx_report_metrics.total_packets++;
+
+    if (conn.last_tx_msg_tick)
+        conn.tx_report_metrics.duration_ms += current_tick - conn.last_tx_msg_tick;
+
+    return true;
+}
+
+
 /*
  * Blocking socket writer. This should be called in its own thread
  *
  * Writer will perform the following:
  *  - loop reads data from fd_write_queue and writes it to the socket
  */
-void
-UDPTransport::fd_writer()
-{
-
-  logger->Log("Starting transport writer thread");
-
-  while (not stop) {
-    auto cd = fd_write_queue.block_pop();
-
-    if (cd) {
-      if ((dequeue_data_map.count(cd->contextId) == 0 || dequeue_data_map[cd->contextId].count(cd->streamId) == 0)
-          || remote_contexts.count(cd->contextId) == 0) {
-        // Drop/ignore connection data since the connection or stream no
-        // longer exists
-        continue;
-      }
-
-      auto& r = remote_contexts.at(cd->contextId);
-
-      int numSent = sendto(fd,
-                           cd.value().data.data(),
-                           cd.value().data.size(),
-                           0 /*flags*/,
-                           (struct sockaddr*)&r.addr,
-                           sizeof(sockaddr_in));
-
-      if (numSent < 0) {
-        logger->error << "Error sending on UDP socket: " << strerror(errno)
-                      << std::flush;
-
-        break;
-
-      } else if (numSent != (int)cd.value().data.size()) {
-        continue;
-      }
+void UDPTransport::fd_writer() {
+    timeval to;
+    to.tv_usec = 1000;
+    to.tv_sec = 0;
+
+    logger->Log("Starting transport writer thread");
+
+    bool sent_data = false;
+    int all_empty_count = 0;
+
+    while (not stop) {
+        sent_data = false;
+
+
+        std::unique_lock<std::mutex> lock(_connections_mutex);
```

Comment:
It's unlocked because close() locks as well. It needs to because close can be called by another thread as well.   This removes deadlock in that case, which I was hitting. :) 

I'll move the unique lock out of the while loop and then lock on each iteration. This is basically protecting connections map over all threads. 

Unlock is before sent data and I added lock after the select. Thanks for the feedback on that.